### PR TITLE
[Frontend] Multi-Modality Support for Loading Local Image Files

### LINKED ISSUE #8730: [Usage]: multimodal large models  load local image files
### Your current environment

PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: CentOS Linux release 7.9.2009 (Core) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: Could not collect
CMake version: version 3.30.1
Libc version: glibc-2.17

Python version: 3.11.9 (main, Apr 19 2024, 16:48:06) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.105.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 12.0.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 545.23.08
cuDNN version: Probably one of the following:
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn.so.8.9.1
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.9.1
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.9.1
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.9.1
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.9.1
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.9.1
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.9.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                12
On-line CPU(s) list:   0-11
Thread(s) per core:    2
Core(s) per socket:    6
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Platinum 8269CY CPU @ 2.50GHz
Stepping:              7
CPU MHz:               2499.976
BogoMIPS:              4999.95
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              36608K
NUMA node0 CPU(s):     0-11
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc eagerfpu pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single rsb_ctxsw fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 arat avx512_vnni

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-ml-py==12.555.43
[pip3] nvidia-nccl-cu12==2.20.5
[pip3] nvidia-nvjitlink-cu12==12.5.82
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] optree==0.12.1
[pip3] pyzmq==26.0.3
[pip3] torch==2.4.0
[pip3] torchvision==0.19.0
[pip3] transformers==4.43.3
[pip3] triton==3.0.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.1.3.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.1.105                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.0.2.54                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.2.106               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.4.5.107               pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.1.0.106               pypi_0    pypi
[conda] nvidia-ml-py              12.555.43                pypi_0    pypi
[conda] nvidia-nccl-cu12          2.20.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.5.82                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.1.105                 pypi_0    pypi
[conda] optree                    0.12.1                   pypi_0    pypi
[conda] pyzmq                     26.0.3                   pypi_0    pypi
[conda] torch                     2.4.0                    pypi_0    pypi
[conda] torchvision               0.19.0                   pypi_0    pypi
[conda] transformers              4.43.3                   pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.6.1.post2@9ba0817ff1eb514f51cc6de9cb8e16c98d6ee44f
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
GPU0	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	0-11	0		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

### How would you like to use vllm

I have deployed multimodal large models via vllm docker containers and my images are on object storage and have been mounted in the container. To load the image faster, I want to read the image mounted locally.
But I found that only http and bast64 encoded input images are supported here. https://github.com/vllm-project/vllm/blob/9b8c8ba1198cbcd311d28b7647f0f8d5dcdc9212/vllm/multimodal/utils.py#L33
Will you support local image files? the code looks like this
```python
elif os.path.isfile(image_url):
    with open(image_url, 'rb') as f:
        image = Image.open(f).convert(image_mode)
```
if not, how do I rebuild my mirror with minimal changes?

Thanks.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### LINKED ISSUE #9742: [Feature]: OpenAI supply image path
### 🚀 The feature, motivation and pitch

Hi! I need a low-latency vision API endpoint. The current vLLM endpoint takes about 200ms (100ms for smaller images) to process a base64 image (provided via `image_url` as base64). I need to cut that number down to almost zero

(I checked it against cached web URL images which have almost no latency)

Would it be possible to implement that one can also supply an absolute path to an image? All my images are local anyways.

### Alternatives

The alternative would be to somehow optimize the base64 parsing, but I don't know if that's possible

### Additional context

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by chaunceyjiang**
FIX #8730
FIX #9742

Multi-Modality Support for Loading Local image Files

### CODE REVIEW

**Comment by DarkLight1337** [comment]
For security reasons, let's require a CLI flag to enable this behavior.

**Comment by chaunceyjiang** [comment]
> For security reasons, let's require a CLI flag to enable loading local image files.

@DarkLight1337  What do you reckon would be a good name for a CLI flag?



How about ` --trust-local-files` ?  
Should this flag be added to the `vllm serve` command?

**Comment by DarkLight1337** [comment]
> > For security reasons, let's require a CLI flag to enable loading local image files.
> 
> @DarkLight1337 What do you reckon would be a good name for a CLI flag?
> 
> How about ` --trust-local-files` ? Should this flag be added to the `vllm serve` command?

I think it is too easily confused with `--trust-remote-code` from HF.

Perhaps `--trust-remote-access`? Allowing API requests to read local image files basically does that.

**Comment by cedonley** [comment]
I would suggest that the flag would instead be the allowed path vs. allowing full filesystem access.

For example, --allow-local-media-path=/some/path

**Comment by chaunceyjiang** [comment]
Test Code:

``` python
from openai import OpenAI

# Modify OpenAI's API key and API base to use vLLM's API server.
openai_api_key = "EMPTY"
openai_api_base = "http://localhost:8000/v1"

client = OpenAI(
    # defaults to os.environ.get("OPENAI_API_KEY")
    api_key=openai_api_key,
    base_url=openai_api_base,
)

models = client.models.list()

model = models.data[0].id




# Single-image input inference
def run_single_image() -> None:

    # Use image url in the payload
    # "https://upload.wikimedia.org/wikipedia/commons/d/da/2015_Kaczka_krzy%C5%BCowka_w_wodzie_%28samiec%29.jpg"
    image_url = "file:///root/jxy/vllm-source/vllm/duck.jpg"
    chat_completion_from_url = client.chat.completions.create(
        messages=[{
            "role":
            "user",
            "content": [
                {
                    "type": "text",
                    "text": "What's in this image?"
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": image_url
                    },
                },
            ],
        }],
        model=model,
        max_completion_tokens=64,
    )

    result = chat_completion_from_url.choices[0].message.content
    print("Chat completion output from image url:", result)


def main() -> None:
   run_single_image()


if __name__ == "__main__":
    main()


```

Test 1:
``` shell

# vllm serve microsoft/Phi-3.5-vision-instruct --task generate \
 --trust-remote-code --max-model-len 4096 --limit-mm-per-prompt image=2 \
 --dtype=float16 --enforce-eager \
 --allowed-local-media-path=/root/jxy/vllm-source/vllm


# python run_single_image.py 
Chat completion output from image url:  The image shows a single mallard duck swimming on a body of water. The duck has a distinctive green head, yellow bill, white neck ring, and brown chest. The water appears calm, creating a clear reflection of the duck.
```


``` shell
# vllm serve microsoft/Phi-3.5-vision-instruct --task generate \
    --trust-remote-code --max-model-len 4096 --limit-mm-per-prompt image=2 \
--dtype=float16 --enforce-eager



# python run_single_image.py                                                                   2 ↵
..
openai.BadRequestError: Error code: 400 - {'object': 'error', 'message': "Invalid 'image_url': Cannot load local files without '--allowed-local-media-path'.", 'type': 'BadRequestError', 'param': None, 'code': 400}

```

File: vllm/multimodal/utils.py [hunk 1]
```diff
@@ -18,19 +19,54 @@
...
 
-def _load_image_from_bytes(b: bytes):
+def _load_image_from_bytes(b: bytes) -> Image.Image:
     image = Image.open(BytesIO(b))
     image.load()
     return image
 
 
-def _load_image_from_data_url(image_url: str):
+def _is_subpath(image_path, allowed_local_media_path):
```
Comment (DarkLight1337) [inline vllm/multimodal/utils.py:28]: Please add type annotations for this function.

File: vllm/entrypoints/llm.py [hunk 2]
```diff
@@ -931,7 +937,8 @@ def _run_engine(
                             # Calculate tokens only for RequestOutput
                             assert output.prompt_token_ids is not None
                             total_in_toks += len(output.prompt_token_ids)
-                            in_spd = total_in_toks / pbar.format_dict["elapsed"]
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:934]: Please avoid formatting unrelated code.

File: vllm/multimodal/utils.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import base64
 from functools import lru_cache
 from io import BytesIO
+from os import path
```
Comment (DarkLight1337) [inline vllm/multimodal/utils.py:4]: Usually we use `os.path` to avoid clashing with local variables. Can you `import os` here?

Comment (chaunceyjiang) [reply vllm/multimodal/utils.py]: Done.

**Review by DarkLight1337 — COMMENTED** [review]
The implementation looks good, can you add some tests to verify this?

**Review by DarkLight1337 — APPROVED** [review]
Thanks for implementing this!
